amazon-kinesis-client/src/main/java/software/amazon/kinesis/coordinator/Scheduler.java

```diff
@@ -1051,6 +1055,73 @@ private StreamIdentifier getStreamIdentifier(Optional<String> streamIdentifierSt
+     */
+    private static StreamConfig withStreamArn(
+            @NonNull final StreamConfig streamConfig, @NonNull final Region kinesisRegion) {
+        Validate.isTrue(streamConfig.streamIdentifier().accountIdOptional().isPresent(),
+                "accountId should not be empty");
+        Validate.isTrue(streamConfig.streamIdentifier().streamCreationEpochOptional().isPresent(),
+                "streamCreationEpoch should not be empty");
+
+        log.info("Constructing stream ARN for {} using the Kinesis client's configured region - {}.",
+                streamConfig.streamIdentifier(), kinesisRegion);
+
+        final StreamIdentifier streamIdentifierWithArn = StreamIdentifier.multiStreamInstance(
+                constructStreamArn(
+                        kinesisRegion,
+                        Long.parseLong(streamConfig.streamIdentifier().accountIdOptional().get()),
+                        streamConfig.streamIdentifier().streamName()),
+                streamConfig.streamIdentifier().streamCreationEpochOptional().get());
+
+        return new StreamConfig(
+                streamIdentifierWithArn, streamConfig.initialPositionInStreamExtended(), streamConfig.consumerArn());
+    }
+
+    @RequiredArgsConstructor
+    private class StreamConfigMap extends ConcurrentHashMap<StreamIdentifier, StreamConfig> {
+        /**
+         * If {@link StreamIdentifier#streamArnOptional()} is present for the provided
+         * {@link StreamConfig#streamIdentifier()}, validates that the region in the stream ARN is consistent with the
+         * region that the Kinesis client ({@link RetrievalConfig#kinesisClient()}) is configured with.
+         * <p>
+         * In multi-stream mode, ensures stream ARN is always present by constructing it using the Kinesis client
+         * region when {@link StreamIdentifier#streamArnOptional()} is {@link Optional#empty()}.
+         * <p>
+         * {@inheritDoc}
+         */
+        @Override
+        public StreamConfig put(
+                @NonNull final StreamIdentifier streamIdentifier, @NonNull final StreamConfig streamConfig) {
+            final Region kinesisRegion = getKinesisRegion();
+
+            final Optional<Arn> streamArnOptional = streamConfig.streamIdentifier().streamArnOptional();
+            if (streamArnOptional.isPresent()) {
+                Validate.isTrue(kinesisRegion.id().equals(streamArnOptional.get().region().get()),
+                        "The provided streamARN " + streamArnOptional.get()
+                                + " does not match the Kinesis client's configured region - " + kinesisRegion);
+                return super.put(streamIdentifier, streamConfig);
+            }
+
+            if (isMultiStreamMode) {
+                return super.put(streamIdentifier, withStreamArn(streamConfig, kinesisRegion));
+            }
+
+            return super.put(streamIdentifier, streamConfig);
```

Comment:
`isPresent()` followed by `get` is an anti-pattern.
Instead use `map` or `ifPresent` as appropriate. In this case, something like
```suggestion
            StreamConfig cfg = streamConfig.streamIdentifier().streamArnOptional().map(streamArn -> {
                Validate.isTrue(kinesisRegion.id().equals(streamArn.region().get()),
                        "The provided streamARN " + streamArn +
                                + " does not match the Kinesis client's configured region - " + kinesisRegion);
                return streamConfig;
            }).orElse(isMultiStreamMode ? withStreamArn(streamConfig, kinesisRegion):streamConfig);
            return super.put(streamIdentifier, cfg);

```

Reply:
Updated. Thanks!